### PR TITLE
Homework

### PR DESCRIPTION
**Pull request by Sandilya1795**
New Origin/Homework created, containing the fix to the StatsCommandTest and StopCommandTest.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -101,6 +101,14 @@
         <!--+
             | Test scope
             +-->
+       <!-- https://mvnrepository.com/artifact/org.apache.httpcomponents/httpclient -->
+        <dependency>
```
Comment (Technolords) [inline pom.xml]: Do we need this dependency?

File: src/test/java/net/technolords/micro/RouteTestSupport.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import org.apache.camel.ProducerTemplate;
+import org.apache.camel.impl.DefaultExchange;
+import org.apache.camel.main.Main;
+import org.apache.camel.testng.CamelTestSupport;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testng.annotations.BeforeTest;
+
+
+public class RouteTestSupport extends CamelTestSupport{
```
Comment (Technolords) [inline src/test/java/net/technolords/micro/RouteTestSupport.java:16]: Test if we can do without CamelTestSupport, as it adds nothing?

File: src/test/java/net/technolords/micro/command/StatsCommandTest.java [hunk 3]
```diff
@@ -18,11 +23,13 @@
 import net.technolords.micro.config.ConfigurationManager;
 import net.technolords.micro.registry.MockRegistry;
 
-public class StatsCommandTest {
+public class StatsCommandTest extends RouteTestSupport {
     private static final Logger LOGGER = LoggerFactory.getLogger(StatsCommandTest.class);
     private Main main;
     private ProducerTemplate producerTemplate;
     private String availablePort;
```
Comment (Technolords) [inline src/test/java/net/technolords/micro/command/StatsCommandTest.java]: Move the part and logic of available port into the parent class (i.e. TestSupport)

File: src/test/java/net/technolords/micro/command/StatsCommandTest.java [hunk 4]
```diff
@@ -59,4 +66,17 @@ public void testStatsCommand() throws Exception{
         LOGGER.info("Got html: {}", html);
     }
 
+    @Test
+    public void testStatsCommand2() throws Exception{
```
Comment (Technolords) [inline src/test/java/net/technolords/micro/command/StatsCommandTest.java]: This method is now a duplicate after the merge.

**Review by Technolords — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 1 above)
Comment (Sandilya1795) [reply pom.xml]: Removed the Dependency.

File: src/test/java/net/technolords/micro/RouteTestSupport.java (hunk 2 above)
Comment (Sandilya1795) [reply src/test/java/net/technolords/micro/RouteTestSupport.java:16]: CamelTestSupport is removed as it adds no added value.

File: src/test/java/net/technolords/micro/command/StatsCommandTest.java (hunk 3 above)
Comment (Sandilya1795) [reply src/test/java/net/technolords/micro/command/StatsCommandTest.java]: While trying to move the logic of available port and setUp to Class Route Test Support, I get NPE. I see in debug mode, my reference object value is null. Test Class is extending RouteTestSupport, but in debug I see TestClassObject reference is not receiving the values from Parent Class. Reading NPE...

Comment (Sandilya1795) [reply src/test/java/net/technolords/micro/command/StatsCommandTest.java]: done, push is pending.

File: src/test/java/net/technolords/micro/command/StatsCommandTest.java (hunk 4 above)
Comment (Sandilya1795) [reply src/test/java/net/technolords/micro/command/StatsCommandTest.java]: done.

**Comment by Technolords** [comment]
After review together, it appears that from full build, the beforeSuite/Test breaks the build, as the stopCommand stops the context and the subsequent test cannot use that same context...

suggestions:
- isolate the stop with its own context
- change the order of the tests when they are executed
- refactor testSupport in a factory, where each test gets its own context (but the wiring etc is done/shared in one place)

The last suggestion has my preference

File: src/test/java/net/technolords/micro/RouteTestSupport.java [hunk 5]
```diff
@@ -1,33 +1,54 @@
...
+import org.apache.camel.testng.AvailablePortFinder;
 import org.apache.camel.testng.CamelTestSupport;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import org.testng.annotations.BeforeSuite;
 import org.testng.annotations.BeforeTest;
 
+import java.util.Properties;
 
 public class RouteTestSupport extends CamelTestSupport{
```
Comment (Technolords) [inline src/test/java/net/technolords/micro/RouteTestSupport.java:16]: CamelTestSupport is not needed, as it adds nothing

**Comment by Technolords** [comment]
Had to try something this morning... namely to change BeforeSuite or BeforeTest into BeforeClass and that worked! Later today I will push my changes and merge it. Consider homework done!

**Pull request merged**